### ISSUE TITLE
allow redirecting file descriptors for the current shell

### ISSUE THREAD

**Issue by frangio** [fish-shell/fish-shell#3948]
There's a feature of bash that allows one to set redirections for the current shell. It's used by running [`exec`](http://wiki.bash-hackers.org/commands/builtin/exec) without a command.

>  If only redirections are given, the redirections affect the current shell without executing any program. 

I've searched and as far as I can tell this is not possible with fish.

**Comment by krader1961** [comment]
Correct, fish currently does not support `exec` with only redirections. I thought there was already an open issue for this but I can't find one so leaving this open. Note that implementing this may be a challenge due to how redirections are handled internally.

**Comment by faho** [comment]
>I thought there was already an open issue for this but I can't find one so leaving this open.

@krader1961: I think you're talking about #3303, where this was _part_ of the discussion.

**Comment by septatrix** [comment]
I am not sure how many use cases there are but one I recently needed are file locks:

```bash
exec 4<>/var/lock/mylockfile
# This form is convenient for locking a file without spawning a subprocess.
# The shell opens the lock file for reading and writing as file descriptor 4,
# then flock is used to lock the descriptor.
flock 4
commands-guarded-by-a-lock
# ...
flock --unlock 4
```

Without the ability to open additional fds one instead has to do the following:

```fish
flock /var/lock/mylockfile fish -c "
    commands-guarded-by-a-lock
    # ...
"
```

In this case I reverted to using bash for the script but having this functionality in fish would be nice

**Comment by kac-** [comment]
> I am not sure how many use cases there are but one I recently needed are file locks:

```bash
# Create pipe and start reader.
mkfifo pipe
cat pipe &
# Open pipe for writing.
exec 3>pipe
echo one >&3
echo two >&3
# Close pipe.
exec 3>&-
```
From [stackoverflow](https://stackoverflow.com/a/8436387)

**Comment by ZeusThundr** [comment]
I have been trying to port some scripts over to fish and currently need redirecting file descriptors. Here is the script for anyone interested allows ueberzug image previews in lf - file manager If anyone knows a hack to make this work inside fish please respond.
```
set -gx FIFO_UEBERZUG $CACHE/ueberzug-$fish_pid
set -f CACHE $HOME/.cache/lf

function cleanup --on-signal SIGINT
    exec 3>&-
    rm $FIFO_UEBERZUG
end

mkfifo $FIFO_UEBERZUG
ueberzug layer -s < $FIFO_UEBERZUG &
exec 3> $FIFO_UEBERZUG

if not test -d $CACHE
  mkdir --parents $CACHE
  command lf $argv 3>&-
else
  command lf $argv
end
```

**Comment by kuvaldini** [comment]
workaround

```fish
gdb --pid $fish_pid --batch -ex 'call (int)close(0)'
cat; and echo CAT; or echo XA
```
```
cat: -: Bad file descriptor
cat: closing standard input: Bad file descriptor
XA
```
need zero(allow) in `/proc/sys/kernel/yama/ptrace_scope`, or edit sources of fish

**Comment by kuvaldini** [comment]
+1 who want this functionality, it does not look popular for 9 years, so i do not believe it will be ever implemented. fish is good in its interactive manner, no good way for complex scripting.

link to original msg https://github.com/fish-shell/fish-shell/issues/12752#issue-4431337595

-----------------------

in bash/zsh/sh/ash/ksh that is done via
```
exec 0<&-
```
then `cat` returns immediately with non-zero status.

#### how to manage in fish shell? are there any binding to C dup(), close(), open()?

super useful for non-interactive scripts.

```
exec 5<0  ## duplicate
exec 0<&- ## close stdin
cat       ## exits with error
grep 'aaa' <&5  ## explicitly consume stdin
```

bash is cumbersome and is bound of UBs and more UBs, but that's pretty cool thing, it can do almost everything in so minimal phrases.

**Comment by krobelus** [comment]
we can probably implement `exec 5<0` etc.
Curious what's the use case in practice. In your example, you might as well not run `cat`.

> fish is good in its interactive manner, no good way for complex scripting.

yeah

**Comment by kuvaldini** [comment]
one problem in stdin-stdout scripting is debug, and error handling. i love bash, zsh and fish for their simplest way work with piping, nor ruby,python,perl any other semi-scripting lang does provide such a quick and short prototyping, especially managing tools with cli.
10 lines in C are 2-3 in C++, 10 lines in C++ are 3 in python, 10 lines of py are 3 lines of shell, of course loosing granular control, like exceptions.

so why redirect stdin.

case 1. imaging dev has script.fish of 200 lines.
```
function func; 
   something
     grep expr  
   ## ^^ pipe| between commands losed, stdin leak, 
   ## ^^ usually leads to read stdin from terminal (in interactive mode) 
   ## ^^ i.e. script (grep) halts on read()....
   ## ^^ or tries to read close fd0 and fails
   ## ^^ or if stdin provided grep steals your input
   ## ^^ and those cases are pain to debug. especially for young devs
   ## ^^ these are annoying pitfails of shell scripting 
end
```
there is a mistake, sometimes it can lead grep consume stdin which is input to main. fish will never say what is wrong - the easiest way to protect => consume stdin explicitly like in prev msg. `grep 'aaa' <&5  ## explicitly consume stdin`

case 2. stdout is payload, do not want to use it as default output for any command like `grep` with lost `-q`, instead of explicit redirection for each command to `>&-` (close) or to `/dev/null` => redirect default to somewhere >close, >null, >default.log, etc and use explicit redirection of payload to >&7 (dup of 1) 

these two helped me working with bash dramatically saving time.

P.S. most LLMs say fish is able to `exec<>X`
